File: R/inst/tests/test-elementwise.r
test.ae <- function()
{
    checkEqualsNumeric(ae(3.4, 3.4), 0)
    checkEqualsNumeric(ae(3.4, 4.4), 1.0)
    checkEqualsNumeric(ae(9, 11), 2)    
}

test.ce <- function()
{
    checkEqualsNumeric(ce(c(1,1,1,0,0,0),c(1,1,1,0,0,0)), 0.0)
    checkEqualsNumeric(ce(c(1,1,1,0,0,0),c(1,1,1,1,0,0)), 1/6)
    checkEqualsNumeric(ce(c(1,2,3,4),c(1,2,3,3)), 1/4)
    checkEqualsNumeric(ce(c("cat","dog","bird"),c("cat","dog","fish")), 1/3)
    checkEqualsNumeric(ce(c("cat","dog","bird"),c("caat","doog","biird")), 1.0)
}

test.mae <- function()
{
    checkEqualsNumeric(mae(0:10, 1:11), 1)
    checkEqualsNumeric(mae(seq(0,2,0.5), seq(0,2,0.5)), 0)
    checkEqualsNumeric(mae(1:4, c(1,2,3,5)), 0.25)    
}

test.mse <- function()
{
    checkEqualsNumeric(mse(0:10, 1:11), 1.0)
    checkEqualsNumeric(mse(seq(0,2,0.5), seq(0,2,0.5)), 0.0)
    checkEqualsNumeric(mse(1:4, c(1,2,3,6)), 1.0)
}

test.msle <- function()
{
    checkEqualsNumeric(msle(c(exp(2)-1),c(exp(1)-1)), 1)
    checkEqualsNumeric(msle(seq(0,2,0.5),seq(0,2,0.5)), 0)
    checkEqualsNumeric(msle(c(1,2,3,exp(1)-1),c(1,2,3,exp(2)-1)), 0.25)
}

test.rmse <- function()
{
    checkEqualsNumeric(rmse(0:10, 1:11), 1.0)
    checkEqualsNumeric(rmse(seq(0,2,0.5), seq(0,2,0.5)), 0.0)
    checkEqualsNumeric(rmse(1:4, c(1,2,3,5)), 0.5)
}

test.rmsle <- function()
{
    checkEqualsNumeric(rmsle(c(exp(2)-1),c(exp(1)-1)), 1)
    checkEqualsNumeric(rmsle(seq(0,2,0.5),seq(0,2,0.5)), 0)
    checkEqualsNumeric(rmsle(c(1,2,3,exp(1)-1),c(1,2,3,exp(2)-1)), 0.5)
}

test.se <- function()
{
    checkEqualsNumeric(se(3.4, 3.4), 0)
    checkEqualsNumeric(se(3.4, 4.4), 1.0)
    checkEqualsNumeric(se(9, 11), 4)
}

test.sle <- function()
{
    checkEqualsNumeric(sle(3.4,3.4), 0)
    checkEqualsNumeric(sle(exp(2)-1,exp(1)-1), 1)
}

File: Python/ml_metrics/test/test_elementwise.py
#! /usr/bin/env python2.7

import unittest
import ml_metrics as metrics
import numpy as np

class TestElementwise(unittest.TestCase):

    def test_ae(self):
        self.assertAlmostEqual(metrics.ae(3.4, 3.4), 0)
        self.assertAlmostEqual(metrics.ae(3.4, 4.4), 1.0)
        self.assertAlmostEqual(metrics.ae(9, 11), 2)

    def test_ce(self):
        self.assertAlmostEqual(metrics.ce([1,1,1,0,0,0],
                                          [1,1,1,0,0,0]), 0)
        self.assertAlmostEqual(metrics.ce([1,1,1,0,0,0],
                                          [1,1,1,1,0,0]), 1.0/6)
        self.assertAlmostEqual(metrics.ce([1,2,3,4],
                                          [1,2,3,3]), 0.25)
        self.assertAlmostEqual(metrics.ce(["cat", "dog", "bird"],
                                          ["cat", "dog", "fish"]), 1.0/3)
        self.assertAlmostEqual(metrics.ce(["cat", "dog", "bird"],
                                          ["caat", "doog", "biird"]), 1)
    
    def test_ll(self):
        self.assertAlmostEqual(metrics.ll(1,1), 0)
        self.assertAlmostEqual(metrics.ll(1,0), np.inf)
        self.assertAlmostEqual(metrics.ll(0,1), np.inf)
        self.assertAlmostEqual(metrics.ll(1,0.5), -np.log(0.5))

    def test_logLoss(self):
        self.assertAlmostEqual(metrics.log_loss([1,1,0,0],[1,1,0,0]), 0)
        self.assertAlmostEqual(metrics.log_loss([1,1,0,0],[1,1,1,0]), np.inf)
        self.assertAlmostEqual(metrics.log_loss([1,1,1,0,0,0],[0.5,0.1,0.01,0.9,0.75,0.001]), 1.881797068998267)
        self.assertAlmostEqual(metrics.log_loss(1,0.5), -np.log(0.5))

    def test_mae(self):
        self.assertAlmostEqual(metrics.mae(range(0,11), range(1,12)), 1)
        self.assertAlmostEqual(metrics.mae([0,.5,1,1.5,2], [0,.5,1,1.5,2]), 0)
        self.assertAlmostEqual(metrics.mae(range(1,5), [1,2,3,5]), 0.25)

    def test_mse(self):
        self.assertAlmostEqual(metrics.mse(range(0,11), range(1,12)), 1)
        self.assertAlmostEqual(metrics.mse([0,.5,1,1.5,2], [0,.5,1,1.5,2]), 0)
        self.assertAlmostEqual(metrics.mse(range(1,5), [1,2,3,6]), 1.0)

    def test_msle(self):
        self.assertAlmostEqual(metrics.msle(np.exp(2)-1,np.exp(1)-1), 1)
        self.assertAlmostEqual(metrics.msle([0,.5,1,1.5,2], [0,.5,1,1.5,2]), 0)
        self.assertAlmostEqual(metrics.msle([1,2,3,np.exp(1)-1], [1,2,3,np.exp(2)-1]), 0.25)

    def test_rmse(self):
        self.assertAlmostEqual(metrics.rmse(range(0,11), range(1,12)), 1)
        self.assertAlmostEqual(metrics.rmse([0,.5,1,1.5,2], [0,.5,1,1.5,2]), 0)
        self.assertAlmostEqual(metrics.rmse(range(1,5), [1,2,3,5]), 0.5)

    def test_rmsle(self):
        self.assertAlmostEqual(metrics.rmsle(np.exp(2)-1,np.exp(1)-1), 1)
        self.assertAlmostEqual(metrics.rmsle([0,.5,1,1.5,2], [0,.5,1,1.5,2]), 0)
        self.assertAlmostEqual(metrics.rmsle([1,2,3,np.exp(1)-1], [1,2,3,np.exp(2)-1]), 0.5)

    def test_se(self):
        self.assertAlmostEqual(metrics.se(3.4, 3.4), 0)
        self.assertAlmostEqual(metrics.se(3.4, 4.4), 1.0)
        self.assertAlmostEqual(metrics.se(9, 11), 4)

    def test_sle(self):
        self.assertAlmostEqual(metrics.sle(3.4, 3.4), 0)
        self.assertAlmostEqual(metrics.sle(np.exp(2)-1, np.exp(1)-1), 1.0)

if __name__ == '__main__':
    unittest.main()
